### ISSUE TITLE
6.4 Breaking Chromatic excludeStories check with storyStoreV7

### ISSUE THREAD

**Issue by AndrewLeedham** [storybookjs/storybook#16951]
**Describe the bug**
In Storybook 6.3 we are using the following `excludeStories` expression:
```javascript
import isChromatic from 'chromatic';

export default {
  component: Component,
  excludeStories: isChromatic() ? undefined : /^Chromatic/;
};
```
This no longer works in 6.4 when using the experimental `storyStoreV7` feature. I can only assume it is related to https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#string-literal-titles, is metadata being pulled with static analysis now?

**To Reproduce**
I created a repro here: https://github.com/AndrewLeedham/storybook-exclude-stories-repro/tree/master I can't publish it to Chromatic because the build fails because of this issue.

**System**
```
System:
    OS: macOS 11.2.3
    CPU: (12) x64 Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
  Binaries:
    Node: 14.18.1 - ~/.nvm/versions/node/v14.18.1/bin/node
    Yarn: 3.1.1 - /usr/local/bin/yarn
    npm: 6.14.15 - ~/.nvm/versions/node/v14.18.1/bin/npm
  Browsers:
    Chrome: 96.0.4664.93
    Edge: 96.0.1054.34
    Firefox: 94.0.1
    Safari: 14.0.3
  npmPackages:
    @storybook/addon-actions: ^6.5.0-alpha.1 => 6.5.0-alpha.1 
    @storybook/addon-docs: ^6.5.0-alpha.1 => 6.5.0-alpha.1 
    @storybook/addon-essentials: ^6.5.0-alpha.1 => 6.5.0-alpha.1 
    @storybook/addon-links: ^6.5.0-alpha.1 => 6.5.0-alpha.1 
    @storybook/react: ^6.5.0-alpha.1 => 6.5.0-alpha.1
```

**Comment by shilman** [comment]
@AndrewLeedham Yes, we're doing static analysis on the story files in `storyStoreV7` mode. It's a breaking change, which is why it's behind an opt-in feature flag.

This is an interesting pattern -- basically a workaround for https://github.com/storybookjs/storybook/issues/9209

If we added that feature (API etc still TBD) would that be a suitable workaround for you?

**Comment by AndrewLeedham** [comment]
@shilman Presumably the static analysis is to speed up discovery for the sidebar, so you don't have to process all the stories on load? Curious what fields it looks at, assuming `include/exclude` and `title`?

Being able to hide things from the sidebar seems like it would resolve my issue yes, as Chromatic will still be able to find the stories but it won't clog things up on the actual storybook :)

**Comment by shilman** [comment]
@AndrewLeedham Yes, you can see the test cases here: https://github.com/storybookjs/storybook/blob/next/lib/csf-tools/src/CsfFile.test.ts

For the moment I think the workaround is to not use `storyStoreV7`. Longer term, we'll need to come up with a better solution for this, possibly using tags.

**Comment by AndrewLeedham** [comment]
@shilman Yeah my plan was to not use `storyStoreV7` for now. Thanks :)

**Comment by AndrewLeedham** [comment]
@shilman could another option be a global `excludeStories` option, meaning it could be run as part of `main.js` once, rather than statically analysed per story.

**Comment by xgebi** [comment]
Is there any progress on this issue? I was able to find a workaround but it would be better if there was a documented way.

**Comment by Sidnioulz** [comment]
Workarounds have been developed through https://github.com/storybookjs/storybook/pull/25328 for people who can afford to have statically analysed config for story exclusion. It seems that documentation is lagging behind, and the doc update can be tracked through https://github.com/storybookjs/storybook/issues/29735.

Sadly, this feature cannot be exposed to dynamic analysis as it would require loading all stories at Storybook start. This would reintroduce the performance issues that warranted the architecture change in Storybook 7. I'll go ahead and close this issue. Please reopen it if I've missed anything not related to static analysis.
